pennylane/capture/autograph/transformer.py

```diff
@@ -0,0 +1,227 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+AutoGraph is a source-to-source transformation system for converting imperative code into
+traceable code for compute graph generation. The system is implemented in the Diastatic-Malt
+package (originally from TensorFlow).
+Here, we integrate AutoGraph into PennyLane to improve the UX and allow programmers to use built-in
+Python control flow and other imperative expressions rather than the functional equivalents provided
+by PennyLane.
+"""
+import copy
+import inspect
+
+from malt.core import converter
+from malt.impl.api import PyToPy
+
+import pennylane as qml
+from pennylane.capture.autograph.ag_primitives import AutoGraphError
+
+from . import ag_primitives
+
+
+class PennyLaneTransformer(PyToPy):
+    """A source-to-source transformer to convert imperative style control flow into a function style
+    suitable for tracing."""
+
+    def __init__(self):
+        super().__init__()
+
+        self._extra_locals = None
+
+    def transform(self, obj, user_context):
+        """Launch the transformation process. Typically, this only works on function objects.
+        Here we also allow QNodes to be transformed."""
+
+        # By default, AutoGraph will only convert function or method objects, not arbitrary classes
+        # such as QNode objects. Here we handle them explicitly, but we might need a more general
+        # way to handle these in the future.
+        # We may also need to check how this interacts with other common function decorators.
+        fn = obj
+        if isinstance(obj, qml.QNode):
+            fn = obj.func
+        elif inspect.isfunction(fn) or inspect.ismethod(fn):
+            pass
+        elif callable(obj):
+            # pylint: disable=unnecessary-lambda,unnecessary-lambda-assignment
+            fn = lambda *args, **kwargs: obj(*args, **kwargs)
+        else:
+            raise AutoGraphError(f"Unsupported object for transformation: {type(fn)}")
+
+        new_fn, module, source_map = self.transform_function(fn, user_context)
+        new_obj = new_fn
+
+        if isinstance(obj, qml.QNode):
+            new_obj = copy.copy(obj)
+            new_obj.func = new_fn
+
+        return new_obj, module, source_map
+
+    def get_extra_locals(self):
+        """Here we can provide any extra names that the converted function should have access to.
+        At a minimum we need to provide the module with definitions for AutoGraph primitives."""
+
+        if self._extra_locals is None:
+            extra_locals = super().get_extra_locals()
+            updates = {key: ag_primitives.__dict__[key] for key in ag_primitives.__all__}
+            extra_locals["ag__"].__dict__.update(updates)
```

Comment:
No, there are a lot of extra things in `ag_primitives.__dict__`, whereas in this PR we've only put 2 things in `ag_primitives.__all__` (we'll add a couple more later). 

So in this case we are only overwriting the autograph definition of `if_stmt` and 
`converted_call`, but not other things in `ag_primitives.__dict__` like `__cached__` or `file`.

Probably nothing bad happens if just pass it everything (nothing bad happens in the tests right now), but it seems safer to overwrite autograph's objects very selectively instead of passing it everything in that file.